### PR TITLE
Skia to support Key-event 

### PR DESCRIPTION
**Pull request by chatrapathi06**

### CODE REVIEW

File: skia/BUILD.gn [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    "-Wno-header-hygiene",
+  ]
+
+  include_dirs = [
+    "./",
+    "//react-native/ReactCommon/",
+  ]
+}
+
+rn_cxx_component("RnRNKeyEvent") {
```
Comment (shefayaparwin) [inline skia/BUILD.gn]: Is RNKeyEvent or RnKeyEvent is not enough?, ANy specific reason to have to  **RnRN**KeyEvent

File: skia/BUILD.gn [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    "RNKeyEvent.cpp"
+  ]
+
+  public_configs = [ ":LibRnRNKeyEvent_config" ]
+  public_configs += [ "//build/secondary/folly:folly_config" ]
+  deps = [
+    "//folly",
+    "//folly:async",
+    "//folly:synchronization",
+    "//skia",
```
Comment (shefayaparwin) [inline skia/BUILD.gn:30]: Is SKia is also a dependent module here ?

File: skia/RNKeyEvent.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/*
+* Copyright (C) 1994-2023 OpenTV, Inc. and Nagravision S.A.
+*
+* This source code is licensed uander the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+
+#include <cxxreact/JsArgumentHelpers.h>
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.cpp]: Header file order to be arranged as per alphabetical order

File: skia/RNKeyEvent.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace facebook {
+namespace xplat {
+
+RNKeyEventModule::RNKeyEventModule() {
+  RNS_LOG_DEBUG("  calling constructor  ");
+}
+
+RNKeyEventModule::~RNKeyEventModule() {
+  RNS_LOG_DEBUG("   calling disstructor  ");
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.cpp]: typo error for  "disstructor  "

File: skia/RNKeyEvent.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
+/*
+* Copyright (C) 1994-2023 OpenTV, Inc. and Nagravision S.A.
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.cpp:2]: copyright  till year differ for .cpp & .h file

File: skia/RNKeyEvent.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+#include <cxxreact/CxxModule.h>
+#include "ReactSkia/core_modules/RSkInputEventManager.h"
+#include "ReactSkia/utils/RnsUtils.h"
+
+using namespace std;
+using namespace folly;
+
+namespace facebook {
+namespace xplat {
+using namespace std;
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.h]: using namespace std; repeating here .Earlier one @ Line NO:12

File: skia/RNKeyEvent.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+#pragma once
+#include <cxxreact/CxxModule.h>
+#include "ReactSkia/core_modules/RSkInputEventManager.h"
+#include "ReactSkia/utils/RnsUtils.h"
+
+using namespace std;
+using namespace folly;
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.h]: Is this needed ? using namespace folly; as in code , it is  referred fully folly::

File: skia/RNKeyEvent.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace facebook {
+namespace xplat {
+using namespace std;
+class RNKeyEventModule : public module::CxxModule{
+ private:
+  int  keyEventId_= 0;
+  int callbackId_ = 0;
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.h:18]: Like for "keyEventId_" : 0 is the invalid case, same for "callbackId_ " ?? callbackId_  can be "0" as well

File: skia/RNKeyEvent.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+using namespace std;
+using namespace folly;
+
+namespace facebook {
+namespace xplat {
+using namespace std;
+class RNKeyEventModule : public module::CxxModule{
+ private:
+  int  keyEventId_= 0;
```
Comment (shefayaparwin) [inline skia/RNKeyEvent.h]: is keyEvenID_ to be unsigned int ?

**Review by shefayaparwin — APPROVED** [review]

**Review by prasanna-jeyam — APPROVED** [review]

File: skia/RNKeyEvent.cpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+std::string RNKeyEventModule::getName() {
+  return "RNKeyEvent";
+}
+
+auto RNKeyEventModule::getMethods() -> std::vector<Method> {
+  return {
+      Method(
+          "addListener",
+          [&] (folly::dynamic args) {
+            ++keyEventId_;
```
Comment (munezbn) [inline skia/RNKeyEvent.cpp:39]: You may have to reset keyEventId_ in destructior ?

File: skia/RNKeyEvent.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+  int callbackId_ = 0;
+  void sendEventWithName(std::string eventName, folly::dynamic eventData);
+ public:
+  RNKeyEventModule();
+  ~RNKeyEventModule();
+  auto getMethods() -> std::vector<Method>;
+  std::string getName();
+};
+#ifdef __cplusplus
```
Comment (munezbn) [inline skia/RNKeyEvent.h]: use RNS_EXPORT_MODULE and move this to cpp file

File: skia/RNKeyEvent.cpp [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+#include <cxxreact/Instance.h>
+#include <cxxreact/JsArgumentHelpers.h>
+
+#include "RNKeyEvent.h"
+
+namespace facebook {
+namespace xplat {
+
+RNKeyEventModule::RNKeyEventModule() {
+  RNS_LOG_DEBUG("  calling constructor  ");
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: debug message to be made meaningful else removed

File: skia/RNKeyEvent.cpp [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+RNKeyEventModule::~RNKeyEventModule() {
+  RNS_LOG_DEBUG("   calling distructor  ");
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: debug message to be made meaningful else removed

File: skia/RNKeyEvent.cpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+  return "RNKeyEvent";
+}
+
+auto RNKeyEventModule::getMethods() -> std::vector<Method> {
+  return {
+      Method(
+          "addListener",
+          [&] (folly::dynamic args) {
+            RNS_LOG_DEBUG("addListener args "<< args[0]);
+            if(args[0].asString() == "onKeyDown")
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: if conditions to be wrapped with braces

File: skia/RNKeyEvent.cpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+          }),// end of addListener lambda
+
+      Method(
+          "removeListeners",
+          [&] (folly::dynamic args) {
+            RNS_LOG_DEBUG(" removeListeners args "<< args[0]<<" listenerCount_ "<<listenerCount_);
+            int  removeCount = args[0].asInt();;
+            listenerCount_ = std::max(listenerCount_ - removeCount, 0);
+            if (listenerCount_ == 0) {
+              RNS_LOG_ERROR("calling stop stopObserving");
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: should this log level be debug ?

File: skia/RNKeyEvent.cpp [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+              stopObserving();
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp:45]: should the eventRegister_ be set to false here ?

File: skia/RNKeyEvent.cpp [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            eventName="onKeyDown";
+          }    
+          if(isKeyUpEventRegister_ && keyInput.action_){
+            eventName="onKeyUp";
+          }
+
+          if(isKeyMultipleEventRegister_ && keyInput.repeat_){
+            eventName="onKeyMultiple";
+          }
+          sendEventWithName(eventName,obj);
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: should the eventName be checked before sending event 
if its not empty, which means event is been registered right

File: skia/RNKeyEvent.cpp [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+namespace facebook {
+namespace xplat {
+
+RNKeyEventModule::RNKeyEventModule() {
+  RNS_LOG_DEBUG("  calling constructor  ");
+}
+
+RNKeyEventModule::~RNKeyEventModule() {
+  RNS_LOG_DEBUG("   calling distructor  ");
+  auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: to avoid duplication , can we call stopObserving ?

File: skia/RNKeyEvent.cpp [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    );//lambda for RSkinputEvent manager registation-end
+}
+
+void RNKeyEventModule::stopObserving(){
+  auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
+  if ( !inputEventManager ) {
+    RNS_LOG_ERROR("Unable to get RSkInputEventManager instance ");
+    return;
+  }
+  inputEventManager->removeKeyEventCallback(callbackId_);
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: callbackId_ need to be reset ?

File: skia/RNKeyEvent.cpp [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+          if(isKeyMultipleEventRegister_ && keyInput.repeat_){
+            eventName="onKeyMultiple";
+          }
+          sendEventWithName(eventName,obj);
+        }
+    );//lambda for RSkinputEvent manager registation-end
+}
+
+void RNKeyEventModule::stopObserving(){
+  auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: if calbackID_ is valid , then only we should do this operation ?

File: skia/RNKeyEvent.h [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+#include "ReactSkia/utils/RnsUtils.h"
+
+
+namespace facebook {
+namespace xplat {
+class RNKeyEventModule : public module::CxxModule{
+ private:
+  int  listenerCount_= 0;
+  int callbackId_ = 0;
+  bool isKeyDownEventRegister_{false};
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.h]: should these bool be atomic ? since it will be accessed by two diff threads

File: skia/RNKeyEvent.h [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+  RNKeyEventModule();
+  ~RNKeyEventModule();
+  auto getMethods() -> std::vector<Method>;
+  std::string getName();
+};
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+xplat::module::CxxModule* RNKeyEventModuleCls(void) {
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.h]: definition to go in cpp file
and can we use RNS_EXPORT_MODULE macro already defined for this purpose

File: skia/RNKeyEvent.cpp [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+void RNKeyEventModule::startObserving(){
+    auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
+    if ( !inputEventManager ) {
+      RNS_LOG_ERROR("Unable to get RSkInputEventManager instance ");
+      return;
+    }
+    callbackId_ = inputEventManager->addKeyEventCallback(
+        [&](react::RSkKeyInput keyInput) {
+        // lambda for RSkinputEvent manager registation.-starting
+          folly::dynamic obj = folly::dynamic::object("pressedKey", RNSKeyMap[keyInput.key_])("action",(int)keyInput.action_)("keyCode",(int)keyInput.key_);
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: do we need to map the keycode values?

File: skia/RNKeyEvent.cpp [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+void RNKeyEventModule::stopObserving(){
+  auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
+  if ( !inputEventManager ) {
+    RNS_LOG_ERROR("Unable to get RSkInputEventManager instance ");
+    return;
+  }
+  if(callbackId_ > 0){
+    inputEventManager->removeKeyEventCallback(callbackId_);
+  }
+  else{
```
Comment (munezbn) [inline skia/RNKeyEvent.cpp]: indentation

File: skia/RNKeyEvent.cpp [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+          folly::dynamic eventPlayload = generateEventPayload(keyInput);
+          sendEventWithName(keyInput.action_ == RNS_KEY_Press?"onKeyDown":"onKeyUp",eventPlayload);
+          if(repeatCount_ > 0 && keyInput.action_ == RNS_KEY_Release){
+            folly::dynamic eventPlayload = generateEventPayload(keyInput,repeatCount_);
+            sendEventWithName("onKeyMultiple",eventPlayload);
+            repeatCount_ = 0;//rest back once you send the event.  
+          }
+        }
+    );//lambda for RSkinputEvent manager registation-end
+}
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp:72]: line gap between end f previous function and new function

File: skia/RNKeyEvent.cpp [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+folly::dynamic RNKeyEventModule::generateEventPayload(react::RSkKeyInput keyInput,unsigned int repeatCount){
+  
+  folly::dynamic eventPlayload = folly::dynamic::object();
+  if(repeatCount > 0 ){
+    eventPlayload["repeatcount"] = repeatCount;
+    //TODO character param need to update
+  }
+
+  eventPlayload["keyCode"] = (int) keyInput.key_;
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp]: should we use c++ style of casting

File: skia/RNKeyEvent.cpp [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+  eventPlayload["pressedKey"] = (std::string) RNSKeyMap[keyInput.key_];
+  return eventPlayload;
+}
+void RNKeyEventModule::stopObserving(){
+  auto inputEventManager = react::RSkInputEventManager::getInputKeyEventManager();
+  if ( !inputEventManager ) {
+    RNS_LOG_ERROR("Unable to get RSkInputEventManager instance ");
+    return;
+  }
+  if(callbackId_ > 0){
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp:88]: this check probably can be done at beginning itself to be call to get inputEventManager

File: skia/RNKeyEvent.cpp [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+    inputEventManager->removeKeyEventCallback(callbackId_);
```
Comment (pprabhu-ng) [inline skia/RNKeyEvent.cpp:94]: callbackID_ to be reset

**Review by munezbn — APPROVED** [review]

**Review by pprabhu-ng — APPROVED** [review]

**Pull request merged**